src/state/slices/portfolioSlice/portfolioSlice.ts

```diff
@@ -94,45 +144,78 @@ export const accountToPortfolio: AccountToPortfolio = args => {
   const portfolio: Portfolio = cloneDeep(initialState)
 
   Object.entries(args).forEach(([_xpubOrAccount, account]) => {
-    const { chain } = account
+    const { chain, pubkey, caip2 } = account
+    const accountSpecifier = `${caip2}:${toLower(pubkey)}`
+
     switch (chain) {
       case ChainTypes.Ethereum: {
         const ethAccount = account as chainAdapters.Account<ChainTypes.Ethereum>
         const { caip2, caip19 } = account
         const CAIP10 = caip10.toCAIP10({ caip2, account: _xpubOrAccount })
+        portfolio.accountBalances.ids.push(accountSpecifier)
+        portfolio.accountSpecifiers.ids.push(accountSpecifier)
 
-        portfolio.accounts.byId[CAIP10] = []
-        portfolio.accounts.byId[CAIP10].push(caip19)
-        portfolio.accounts.ids.push(CAIP10)
+        portfolio.accounts.byId[accountSpecifier] = []
+        portfolio.accounts.byId[accountSpecifier].push(caip19)
+        portfolio.accounts.ids.push(accountSpecifier)
 
-        portfolio.balances.byId[caip19] = ethAccount.balance
-        portfolio.balances.ids.push(caip19)
+        portfolio.assetBalances.byId[caip19] = ethAccount.balance
+        portfolio.assetBalances.ids.push(caip19)
+
+        portfolio.accountBalances.byId[accountSpecifier] = {
+          [caip19]: ethAccount.balance
+        }
+
+        portfolio.accountSpecifiers.byId[accountSpecifier] = [CAIP10]
 
         ethAccount.chainSpecific.tokens?.forEach(token => {
           portfolio.accounts.byId[CAIP10].push(token.caip19)
-          portfolio.balances.ids.push(token.caip19)
-          portfolio.balances.byId[token.caip19] = token.balance
+          portfolio.assetBalances.ids.push(token.caip19)
+          portfolio.assetBalances.byId[token.caip19] = token.balance
+
+          portfolio.accountBalances.byId[accountSpecifier] = {
+            ...portfolio.accountBalances.byId[accountSpecifier],
+            [token.caip19]: token.balance
+          }
         })
         break
       }
       case ChainTypes.Bitcoin: {
         const btcAccount = account as chainAdapters.Account<ChainTypes.Bitcoin>
         const { caip2, caip19 } = account
-        const addresses = btcAccount.chainSpecific.addresses ?? []
-        if (addresses.length) portfolio.balances.ids.push(caip19)
+        const addresses =
+          btcAccount.chainSpecific.addresses?.filter(addy => bnOrZero(addy.balance).gt(0)) ?? []
```

Comment:
@0xdef1cafe Should we filter addresses to only those that have a balance? It made sense when we were writing it, since they. don't add to the balance of the portfolio. However, if we are reading receive addresses from state, then we'll need a way to keep the next receive address in the array.

Reply:
We don't want to filter, because we need CAIP10s with zero balance for tx history mapping.